### ISSUE TITLE
Fable.Elmish: Exception in update after Cmd.OfAsync.* is silently ignored

### ISSUE THREAD

**Issue by jindraivanek** [fable-compiler/fable-elmish#287]
### Description

Exception in `update` after Cmd.OfAsync.* is silently ignored and Elmish loop stops. Problem is in Fable.Elmish, works correctly in dotnet.

### Repro code
Simple repro is here: https://github.com/jindraivanek/repro-elmish-ignored-exception/blob/main/Program.fs

### Expected and actual results

Exception should be raised.

### Possible fix

I was able to fix it by not using `Timer.delay` here, just `Async.StartImmediate`: https://github.com/elmish/elmish/blob/236787ea8b12d652ac71398095611bacdbc234d2/src/cmd.fs#L123 

Is `Timer.delay` really needed here?

### Related information 

* elmish version: 4.1.0
* fable-compiler version: 4.18.0
* fable-core version: 3.7.1
* Operating system: Ubuntu 22.4

**Comment by et1975** [comment]
The `delay` is there to simulate `Async.start` behaviour, which unlike `Async.StartImmediate` schedules the computation on the pool instead of running it on the current thread. 
Regardless the exception handler is inside the code block that shouldn't get affected by w/ or not it's called from the timer or directly.
~~@alfonsogarciacaro any idea why [`Async.Catch`](https://github.com/elmish/elmish/blob/v4.x/src/cmd.fs#L86) would behave differently here?~~

**Comment by MangelMaxime** [comment]
If there is a bug / question with Fable, it is better to open it on Fable repository so all Fable maintainers see it.

This is because Alfonso is not active anymore and I am not sure if others maintainers watch Elmish repository for notifications.

pinging @ncave if they have an idea regarding the question above

**Comment by ncave** [comment]
@MangelMaxime Yes, I agree if there is a general issue with Async, it should be opened on the Fable repo.
I can't say I'm familiar with the Async implementation, but perhaps if we get a simpler repro, we could see what's happening in the generated JS code.

**Comment by jindraivanek** [comment]
The exception happens in `update` that is part of `dispatch` and is not guarded by `Async.Catch`, so problem isn't impl of `Async.catch`.

Interesting thing: If I use the same code that is used for Fable in .NET. `Timer.delay 1 (fun _ -> Async.StartImmediate x)`, exception is also ignored. Looks like this bug/feature is there for long time: https://stackoverflow.com/questions/3749785/how-do-i-get-the-exception-that-happens-in-timer-elapsed-event

On other hand `Timer.delay 1 (fun _ -> Async.Start x)` raises exception in .NET, bot not in Fable. That makes sense considering Fable emits warning `Async.Start will behave as StartImmediate`.

So, question is if this is actually bug in Fable `Timer` impl or not.

Simple repro code of `Timer` swallowing exception in async:

```fsharp
module Timer
open System.Timers
let delay interval callback =
    let t = new Timer(float interval, AutoReset = false)
    t.Elapsed.Add callback
    t.Enabled <- true
    t.Start()

delay 1 (fun _ -> async { failwith "BOOM" } |> Async.StartImmediate)
```

Should I open an issue in Fable about this? 

Other option would be workaround in Elmish by `Async.catch` whole async in https://github.com/elmish/elmish/blob/v4.x/src/cmd.fs#L85-L90 and other `OfAsync` functions.

**Comment by ncave** [comment]
@jindraivanek 
> If I use the same code that is used for Fable in .NET, exception is also ignored.
> So, question is if this is actually bug in Fable Timer impl or not.

If the behavior is the same as in .NET, probably not.

Here is the .NET Timer behavior as documented:
> The [Timer](https://learn.microsoft.com/en-us/dotnet/api/system.timers.timer?view=net-8.0) component catches and suppresses all exceptions thrown by event handlers for the [Elapsed](https://learn.microsoft.com/en-us/dotnet/api/system.timers.timer.elapsed?view=net-8.0#system-timers-timer-elapsed) event. This behavior is subject to change in future releases of the .NET Framework. Note, however, that this is not true of event handlers that execute asynchronously

**Comment by et1975** [comment]
> The exception happens in `update` that is part of `dispatch` and is not guarded by `Async.Catch`, so problem isn't impl of `Async.catch`.

Right, sorry, misread the situation. So, it's not in the try/catch of the async block, but is still on the callback thread, which in .NET is probably a pool thread and anything running on it cannot throw.
However I just noticed between v3 and v4 we've lost try/catch around the `update` call, @kspeakman do you recall why?

**Comment by kspeakman** [comment]
I believe I was thinking:
- side effect exceptions are already handled by [`Cmd.exec`](https://github.com/elmish/elmish/blob/v4.x/src/program.fs#L190) (synchronous ones anyway)
- if a decision in `update` can throw, the dev probably wants to handle that logical outcome

I'm not against putting it back in if it helps the dev experience. IOW, if having Elmish handle the error makes it more traceable than an unhandled exception stack trace. I don't think it does, since Elmish will tell you what message failed, but the unhandled stack trace should point to the update line that failed.

It's trivial for a dev to implement themselves, should their situation call for it.

```fsharp
let updateTryCatch msg model =
    try
        update msg model
    catch ex ->
        // your desired behavior on exception goes here
        model, Cmd.none

Program.mkProgram init updateTryCatch view
...
|> Program.run
```

**Comment by MangelMaxime** [comment]
I also got tricked by the issue reported in this issue.

I don't think asking the user to `try ... catch` is own update functions is a good idea. This is because, it will probably never be done because nothing tells you to do it.

I think having the old behaviour is better, because it helps user identify problems and also allows the Elmish program to keep working. I agree that the exceptions should normally be handled in the `Cmd` module, but mistakes happens.

**Comment by et1975** [comment]
I'd prefer to reinstate it:
- this feels like a regression since the previous version did handle it
- with F# Core throwing left and right the language itself does not facilitate exception-free code
- while our own handling has always been the last line of defence, the guidance has been - the user absolutely should guard against exceptions, the fallback reporting via Program's `onError` goes a long away in detecting it
- tooling hookup via `onError` - something like `withDebuger` channels the errors correctly and asking the user to re-do it on the side would make it needlessly complicated
- on .NET depending on which thread calls the dispatch this may end up crashing the process

A PR would be welcome.

**Comment by kspeakman** [comment]
My thought was that the exception will inform the dev that a try catch is needed if they didn't already know. And that the user experience is bad in production for Elmish's generic handler. To the user, it looks like their actions were ignored. At least a dev-implemented generic handler can update the model to inform the user there was an error. Elmish doesn't know enough details of the model to do that.

Really the same can be said for Cmd.exec's exception handling. Async exceptions still must be handled by the dev.

No worries from me with adding it back if that's the consensus. But I think this should be dev responsibility or we should have a more comprehensive approach to error handling.

**Comment by MangelMaxime** [comment]
> My thought was that the exception will inform the dev that a try catch is needed if they didn't already know.

The problem here and from what I remembered in my case, was that no exception was reported and I had nothing in the console either. So, I was not informed of the problem.

**Comment by kspeakman** [comment]
> > My thought was that the exception will inform the dev that a try catch is needed if they didn't already know.
> 
> The problem here and from what I remembered in my case, was that no exception was reported and I had nothing in the console either. So, I was not informed of the problem.

Ah, yeah that's not good.

**Comment by jindraivanek** [comment]
> The problem here and from what I remembered in my case, was that no exception was reported and I had nothing in the console either. So, I was not informed of the problem.

Yes, that is exactly the main issue.

**Comment by et1975** [comment]
The fix is released in 4.2.0
